### PR TITLE
docs: Troubleshooting guide

### PR DESCRIPTION
**Pull request by Arvindthiru**
### Description of your changes

Fixes #

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: docs/troubleshooting/README.md [hunk 1]
```diff
@@ -1 +1,114 @@
...
+
+### How to find the latest ClusterResourceSnapshot resource?
+
+Replace **{CRPName}** in the command below with name of CRP
+
+```
+$ kubectl get clusterresourcesnapshot -l kubernetes-fleet.io/is-latest-snapshot=true,kubernetes-fleet.io/parent-CRP={CRPName} -o YAML
+```
+
+### How can I debug when my CRP doesn't pick up the latest change?
```
Comment (Arvindthiru) [inline docs/troubleshooting/README.md:114]: Still figuring out the best way to answer this question.

File: docs/troubleshooting/README.md [hunk 2]
```diff
@@ -1 +1,114 @@
...
+
+### How to find the latest ClusterSchedulingSnapshot resource?
+
+We need to have ClusterResourcePlacement's name **{CRPName}**, replace **{CRPName}** in the command below,
+
+```
+$ kubectl get clusterschedulingpolicysnapshot -l kubernetes-fleet.io/is-latest-snapshot=true,kubernetes-fleet.io/parent-CRP={CRPName}
+```
+
+### How can I debug when my CRP status is ClusterResourcePlacementSynchronized condition status is set to "False"?
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:29]: could you please list when the sync will be set as "false" first?

 providing the troubleshooting steps is helping them to validate the case and find the action item.

Here, you only provide the validation part, but not the action item.

File: docs/troubleshooting/README.md [hunk 3]
```diff
@@ -1 +1,114 @@
...
+$ kubectl get work -n fleet-member-{clusterName} -l kubernetes-fleet.io/parent-CRP={CRPName},kubernetes-fleet.io/parent-resource-binding={CRBName} -o YAML
+```
+
+### How can I debug when some clusters are not selected as expected?
+
+Check the status of the **ClusterSchedulingPolicySnapshot** to determine which clusters were selected along with the reason
+
+### How can I debug when a selected cluster does not have the expected resources on it?
+
+We need to take a look at the **placementStatuses** section in CRP status for that particular cluster in ClusterResourcePlacement's status. In **placementStatuses** we would find **failedPlacements** which should have the reason
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:104]: list possible reasons
1. latest ClusterResourceSnapshot has not been created.
2. the sync has not been finished.
3. the apply has not been finished.
4. the apply fails.

please provide validation steps for each case and its action item.

File: docs/troubleshooting/README.md [hunk 4]
```diff
@@ -1 +1,114 @@
-TBA.
\ No newline at end of file
+# Troubleshooting guide
```
Comment (michaelawyu) [inline docs/troubleshooting/README.md:1]: Hi Arvind! Let's talk more about this in our sync today; thanks for the doc work 🙏

File: docs/troubleshooting/README.md [hunk 5]
```diff
@@ -1 +1,611 @@
...
+
+**ClusterResourcePlacementSynchronized** condition status is set to **"False"** if the following occurs, the work is not created/updated for a new **ClusterResourceSnapshot**, **ClusterResourceBinding** for a given cluster.
+
+In the **ClusterResourcePlacement** status section check to see which **placementStatuses** also has WorkSynchronized status set to **false**.
+
+From the **placementStatus** we can get the **clusterName** and then check the **fleet-member-{clusterName}** namespace to see if a work objects exists/updated in this case it won't as **WorkSynchronized** has failed.
+
+We need to find the corresponding **ClusterResourceBinding** for our **ClusterResourcePlacement** which should have the status of **work** create/update.
+
+A common case where this could happen is user input for the **rollingUpdate** config it too strict for rolling update strategy.
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:191]:
```suggestion
A common case where this could happen is user input for the **rollingUpdate** config it too strict for rolling update strategy.

Please check the **rollingUpdate** strategy to make sure the maxUnavailable and maxSurge meets your expectations.
```

File: docs/troubleshooting/README.md [hunk 6]
```diff
@@ -1 +1,611 @@
...
+  - kind: Namespace
+    name: test-ns
+    version: v1
+```
+
+we see that **ClusterResourcePlacementSynchronized** is set to false and the message reads **"Works need to be synchronized on the hub cluster or there are still manifests pending to be processed by the 2 member clusters"**. We have this situation cause **rollingUpdate** input was not specified by the user and hence by default,
+
+**maxUnavailable** is set to 1 and **maxSurge** is set to 1.
+
+Meaning after the CRP was created first two **ClusterResourceBindings** were created and since the namespace didn't exist on the hub cluster we did not have to create work and **ClusterResourcePlacementSynchronized** was set to **True**.
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:366]: This's wrong, i remember we still create the work, but the manifest is empty.

File: docs/troubleshooting/README.md [hunk 7]
```diff
@@ -1 +1,611 @@
...
+      type: ResourceApplied
+  selectedResources:
+  - kind: Namespace
+    name: test-ns
+    version: v1
+```
+
+we see that **ClusterResourcePlacementSynchronized** is set to false and the message reads **"Works need to be synchronized on the hub cluster or there are still manifests pending to be processed by the 2 member clusters"**. We have this situation cause **rollingUpdate** input was not specified by the user and hence by default,
+
+**maxUnavailable** is set to 1 and **maxSurge** is set to 1.
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:364]:
```suggestion
**maxUnavailable** is set to 25% * 3 (desired number) and rounded to 1 and **maxSurge** is set to 25% * 3 (desired number) and rounded to 1.
```

File: docs/troubleshooting/README.md [hunk 8]
```diff
@@ -1 +1,611 @@
...
+
+Meaning after the CRP was created first two **ClusterResourceBindings** were created and since the namespace didn't exist on the hub cluster we did not have to create work and **ClusterResourcePlacementSynchronized** was set to **True**.
+
+But once we create the **test-ns** namespace on the hub the rollout controller tries to pick the two **ClusterResourceBindings** to update, but we have **maxUnavailable** set to 1 which is already the case since we have one missing member cluster now if when the rollout controller tries to roll out the updated **ClusterResourceBindings** and even if one of them fails to apply we break the criteria of the **rollout config** since **maxUnavailable** is set 1.
+
+The solution to this particular case is to manually set maxUnavailable to a higher value than 2 to avoid this scenario.
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:370]:
```suggestion
The solution to this particular case is to manually set maxUnavailable to a higher value than 2 to loose the rolling update configuration or join the third member cluster.
```

File: docs/troubleshooting/README.md [hunk 9]
```diff
@@ -1 +1,611 @@
...
+kubectl get clusterresourcesnapshot -l kubernetes-fleet.io/is-latest-snapshot=true,kubernetes-fleet.io/parent-CRP={CRPName} -o YAML
+```
+
+### How can I debug when my CRP ClusterResourcePlacementApplied condition is set to "False"?
+
+In the **ClusterResourcePlacement** status section check to see which **placementStatuses** also has ResourceApplied status set to false.
+
+From the **placementStatuses** we can get the **clusterName** and then use it to find the work object associated with the member cluster in the **fleet-member-{ClusterName}** namespace in the hub cluster and check its status to figure out what's wrong.
+
+example, in this case the **CRP** is trying to propagate a namespace which contains a deployment to two member clusters, but the namespace already exists on one member cluster called **kind-cluster-1**
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:433]:
```suggestion
For example, in this case the **CRP** is trying to propagate a namespace which contains a deployment to two member clusters, but the namespace already exists on one member cluster called **kind-cluster-1**
```

File: docs/troubleshooting/README.md [hunk 10]
```diff
@@ -1 +1,611 @@
...
+    namespace: test-ns
+    version: v1
+  - kind: Namespace
+    name: test-ns
+    version: v1
+```
+
+in the **placementStatuses** section of the **CRP status** for kind-cluster-1 in **failedPlacements** we get a clear message as to why the resource failed to apply on the member cluster.
+
+At times, we might need more information in that case please take a look at the work object
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:544]: provide the link how to find the corresponding work?

File: docs/troubleshooting/README.md [hunk 11]
```diff
@@ -1 +1,611 @@
...
+        type: Applied
+      identifier:
+        kind: Namespace
+        name: test-ns
+        ordinal: 1
+        resource: namespaces
+        version: v1
+```
+
+from looking at the **work status** and specifically the **manifestConditions** section we could see that the namespace could not be applied but the deployment within the namespace got propagated from hub to the member cluster correctly. In this case to solve this issue maybe delete the existing namespace on the member cluster but that's upto the user to decide since the namespace could already contain resources within it.
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:586]:
```suggestion
From looking at the **work status** and specifically the **manifestConditions** section we could see that the namespace could not be applied but the deployment within the namespace got propagated from hub to the member cluster correctly. In this case to solve this issue maybe delete the existing namespace on the member cluster but that's upto the user to decide since the namespace could already contain resources within it.
```

File: docs/troubleshooting/README.md [hunk 12]
```diff
@@ -1 +1,611 @@
...
+Check the status of the **ClusterSchedulingPolicySnapshot** to determine which clusters were selected along with the reason.
+
+### How can I debug when a selected cluster does not have the expected resources on it/ if CRP doesn't pick up the latest changes?
+
+Please check the following cases,
+- check to see if **ClusterResourcePlacementSynchronized** condition in CRP status is set to **True** or **False**
+- If it's set to **false** check this [question](#how-can-i-debug-when-my-crp-status-is-clusterresourceplacementsynchronized-condition-status-is-set-to--false--)
+- If it's set to **True**,
+  - check to see if **ClusterResourcePlacementApplied** condition is set to **Unknown**, **False** or **True**
+  - if it's set to **Unknown** please wait as the resources are still being applied to the member clusters (if it's stuck in unknown state please raise a github issue as it's an unexpected behavior)
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:607]:
```suggestion
  - if it's set to **Unknown** please wait as the resources are still being applied to the member clusters (if it's stuck in unknown state for a while, please raise a github issue as it's an unexpected behavior)
```

File: docs/troubleshooting/README.md [hunk 13]
```diff
@@ -1 +1,611 @@
...
+    selected: true
+  - clusterName: kind-cluster-2
+    reason: 'ClusterUnschedulable, none of the nonempty required cluster affinity
+      term (total number: 1) is matched'
+    selected: false
+```
+
+The solution here is to add the **env:prod** label to the member cluster resource for **kind-cluster-2** as well so that the scheduler can pick the cluster to propagate resources.
+
+### How to verify the latest ClusterSchedulingPolicySnapshot for a CRP?
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:170]: Could you please move all of the validating/checking questions to the bottom? they're more like debugging tools/apis.

How to verify the latest ClusterSchedulingPolicySnapshot for a CRP?
How to find the latest ClusterResourceBinding resource?
...

File: docs/troubleshooting/README.md [hunk 14]
```diff
@@ -278,11 +273,92 @@ status:
       type: ResourceApplied
 ```
 
-Since the resource **test-ns** namespace never existed on the hub cluster **ClusterResourcePlacementApplied** is set to True but **ClusterResourcePlacementScheduled** is set to false since the spec wants to pick 3 clusters but the **scheduler** can only schedule to two available joined clusters.
+Given that the resource **test-ns** namespace never existed on the hub cluster, the `ClusterResourcePlacement` status reflects the following:
+- `ClusterResourcePlacementApplied` is set to `true`
+- `ClusterResourcePlacementScheduled` is set to `false`, as the specified policy aims to pick three clusters, but the scheduler can only accommodate placement in two currently available and joined clusters.
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:278]:
```suggestion
- `ClusterResourcePlacementScheduled` is set to `false`, as the specified policy aims to pick three clusters, but the scheduler can only accommodate placement in two currently available and joined clusters.
- `ClusterResourcePlacementSynchronized` is set to `true`
- `ClusterResourcePlacementApplied` is set to `true`
```

File: docs/troubleshooting/README.md [hunk 15]
```diff
@@ -278,11 +273,92 @@ status:
...
+    type: Bound
+  - lastTransitionTime: "2023-12-05T04:17:50Z"
+    message: ""
+    observedGeneration: 2
+    reason: AllWorkHasBeenApplied
+    status: "True"
+    type: Applied
+```
+
+Upon inspecting the `ClusterResourceBinding` spec, we observe that the `resourceSnapshotName` matches the latest `ClusterResourceSnapshot` name. Additionally, both the `Bound` and `Applied` conditions are set to True.
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:357]: can we delete these statements? too much for the reader to follow. You just want to demonstrate why the synchronize is false after changing the resources.

These contents are unrelated.

File: docs/troubleshooting/README.md [hunk 16]
```diff
@@ -359,81 +435,119 @@ status:
...
-
-But once we create the **test-ns** namespace on the hub the rollout controller tries to pick the two **ClusterResourceBindings** to update, but we have **maxUnavailable** set to 1 which is already the case since we have one missing member cluster now if when the rollout controller tries to roll out the updated **ClusterResourceBindings** and even if one of them fails to apply we break the criteria of the **rollout config** since **maxUnavailable** is set 1.
-
-The solution to this particular case is to manually set maxUnavailable to a higher value than 2 to avoid this scenario.
-
-### How to find the latest ClusterResourceBinding resource?
-
-We need to have ClusterResourcePlacement's name **{CRPName}**, replace **{CRPName}** in the command below. The command below lists all **ClusterResourceBindings** associated with **ClusterResourcePlacement**
+Let's check the latest `ClusterResourceSnapshot`, please check this [section](#how-to-find-the-latest-clusterresourcesnapshot-resource) for more details,
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:440]:
```suggestion
Let's check the latest `ClusterResourceSnapshot`. Please refer to this [section](#how-to-find-the-latest-clusterresourcesnapshot-resource) to learn how to get the latest `ClusterResourceSnapshot`.
```

File: docs/troubleshooting/README.md [hunk 17]
```diff
@@ -359,81 +435,119 @@ status:
...
-
-```
-kubectl get crp test-crp
-NAME       GEN   SCHEDULED   SCHEDULEDGEN   APPLIED   APPLIEDGEN   AGE
-test-crp   1     True        1              True      1            15s
-```
+Upon inspecting `ClusterResourceSnapshot` spec, we observe that the `selectedResources` section now has the namespace `test-ns`.
 
-the **placementStatuses** of the **CRP** above looks like, it has propagated resources to two member clusters and hence has two **ClusterResourceBindings**,
+Let's check the `ClusterResourceBinding` for `kind-cluster-1` to see if it got updated after the namespace `test-ns` was created, please check this [section](#how-to-find-the-latest-clusterresourcebinding-resource) for more details,
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:480]:
```suggestion
Let's check the `ClusterResourceBinding` for `kind-cluster-1` to see if it got updated after the namespace `test-ns` was created. Please check this [section](#how-to-find-the-latest-clusterresourcebinding-resource) to learn how to get the latest `ClusterResourceBinding`.
```

File: docs/troubleshooting/README.md [hunk 18]
```diff
@@ -359,81 +435,119 @@ status:
...
+- Alternatively, you can also join a third member cluster.
 
 ### How can I debug when my CRP ClusterResourcePlacementApplied condition is set to "False"?
 
-In the **ClusterResourcePlacement** status section check to see which **placementStatuses** also has ResourceApplied status set to false.
+#### Investigation steps:
 
-From the **placementStatuses** we can get the **clusterName** and then use it to find the work object associated with the member cluster in the **fleet-member-{ClusterName}** namespace in the hub cluster and check its status to figure out what's wrong.
+1. Check `placementStatuses`: In the `ClusterResourcePlacement` status section, inspect the `placementStatuses` to identify which clusters have the `ResourceApplied` condition set to `false` and note down their `clusterName`.
+2. Locate `Work` Object in Hub Cluster: Use the identified `clusterName` to locate the `Work` object associated with the member cluster.
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:544]:
```suggestion
2. Locate `Work` Object in Hub Cluster: Use the identified `clusterName` to locate the `Work` object associated with the member cluster. Please refer to xxx to learn how to get the `Work`.
```

File: docs/troubleshooting/README.md [hunk 19]
```diff
@@ -583,29 +696,100 @@ At times, we might need more information in that case please take a look at the
...
+### How can I debug when a selected cluster does not have the expected resources on it or if CRP doesn't pick up the latest changes?
+
+Please check the following cases,
+- Check to see if `ClusterResourcePlacementSynchronized` condition in CRP status is set to `true` or `false`.
+- If it's set to `false` check this [question](#how-can-i-debug-when-my-crp-status-is-clusterresourceplacementsynchronized-condition-status-is-set-to--false--).
+- If it's set to `true`,
+  - Check to see if `ClusterResourcePlacementApplied` condition is set to `unknown`, `false` or `true`.
+  - If it's set to `unknown`, please wait as the resources are still being applied to the member cluster (if it's stuck in unknown state for a while, please raise a github issue as it's an unexpected behavior).
+  - If it's set to `false`, check this [question](#how-can-i-debug-when-my-crp-clusterresourceplacementapplied-condition-is-set-to--false--).
+  - If it's set to `true`, check to see if the resource exists on the hub cluster. The `ClusterResourcePlacementApplied` condition is set to `true` if the resource doesn't exist on the hub.
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:717]: - If it's set to `true`, check to see if the resource exists on the hub cluster.~~The `ClusterResourcePlacementApplied` condition is set to `true` if the resource doesn't exist on the hub.~~


The second statement looks confusing.

File: docs/troubleshooting/README.md [hunk 20]
```diff
@@ -165,35 +171,24 @@ status:
...
-In the **ClusterResourcePlacement** status section check to see which **placementStatuses** also has WorkSynchronized status set to **false**.
+#### **Investigation Steps:**
 
-From the **placementStatus** we can get the **clusterName** and then check the **fleet-member-{clusterName}** namespace to see if a work objects exists/updated in this case it won't as **WorkSynchronized** has failed.
+- In the **ClusterResourcePlacement** status section, examine the **placementStatuses** to identify clusters with the `WorkSynchronized` status set to **"False"**.
+- Locate the corresponding **ClusterResourceBinding** for the identified cluster. This resource should indicate the status of the `Work`, whether it was created or updated.
+- A common scenario leading to this issue is the user input for the **rollingUpdate** configuration being too strict. Verify the values for `maxUnavailable` and `maxSurge` to ensure they align with your expectations.
 
-We need to find the corresponding **ClusterResourceBinding** for our **ClusterResourcePlacement** which should have the status of **work** create/update.
+#### **Example Scenario:**
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:187]: sometimes you're using ** and sometimes you are not. Can we make them consistent?

File: docs/troubleshooting/README.md [hunk 21]
```diff
@@ -1 +1,717 @@
...
+
+Replace `{CRPName}` in the command below with name of `ClusterResourcePlacement`,
+
+```
+kubectl get clusterresourcesnapshot -l kubernetes-fleet.io/is-latest-snapshot=true,kubernetes-fleet.io/parent-CRP={CRPName} -o YAML
+```
+
+## How and where to find the correct Work resource?
+
+We need to have the member cluster's namespace which follow this format `fleet-member-{clusterName}`, `ClusterResourceBinding` name `{CRBName}` and `ClusterResourcePlacement` name `{CRPName}`.
```
Comment (zhiying-lin) [inline docs/troubleshooting/README.md:713]: we don't need the `ClusterResourceBinding` name right? we can still get the work created for the CRP.

**Review by zhiying-lin — APPROVED** [review]
